### PR TITLE
Add SOAP Extension

### PR DESCRIPTION
**Pull request by aemadrid**
Adds another option to render SOAP tags with a custom extension loaded with `Papercraft.extension(soap: Papercraft::SoapComponents)`.

This provides four SOAP tags:

* `soap.Envelope()` => <soap:Envelope/>` (with default `xmlns:soap` attribute since it is required and fixed)
* `soap.Header()` => <soap:Header/>`
* `soap.Body()` => <soap:Body/>`
* `soap.Fault()` => <soap:Fault/>`

For tag attributes with colons we still need the replacement of `__` with `:`. For example `xmlns__soap: 'http://schemas.xmlsoap.org/soap/envelope/'` would translate to
`xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/"`.

### CODE REVIEW

File: lib/papercraft/xml.rb [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 require 'escape_utils'
 require_relative './tags'
+require_relative './soap'
```
Comment (noteflakes) [inline lib/papercraft/xml.rb]: If we're going in the direction of Papercraft including some "standard" extensions, I'd rather they be not required by default. We should prefer that people do something like:

```ruby
require 'papercraft/extensions/soap'

# The extension is automatically loaded
xml = Papercraft.xml { ... }
```

See also my comment about the file location for soap.rb.

File: lib/papercraft/soap.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
+module Papercraft
```
Comment (noteflakes) [inline lib/papercraft/soap.rb]: Extensions bundled with Papercraft should have their own namespace, and live in a separate subdirectory. We can put them in `lib/papercraft/extensions`, and define them under `Papercraft::Extensions`. So the module will become `Papercraft::Extensions::SoapComponents` or something of the kind.

File: lib/papercraft/soap.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  module SoapComponents
+    def Envelope(**props, &blk)
```
Comment (noteflakes) [inline lib/papercraft/soap.rb]: All methods (and modules for that sake) should have docs. I'm using the Yard system of tags for documenting method parameters and return value. Have a look at the docs for the `Renderer` class for example: https://github.com/digital-fabric/papercraft/blob/master/lib/papercraft/renderer.rb

File: test/test_xml.rb [hunk 4]
```diff
@@ -100,6 +100,35 @@ def test_soap_issue_9
     assert_equal(
       '<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"><soap:Body><PosRequest xmlns="http://Some.Site"><Ver1.0><Header><SecretAPIKey>some_secret_key</SecretAPIKey></Header><Transaction><SomeData></SomeData></Transaction></Ver1.0></PosRequest></soap:Body></soap:Envelope>'.gsub('<', '\n<'),
       xml.render.gsub('<', '\n<')
-    )    
+    )
+  end
+
+  def test_soap_extension
```
Comment (noteflakes) [inline test/test_xml.rb]: This test should be moved to a separate file, for example `test_soap_extension.rb`.

**Review by noteflakes — CHANGES_REQUESTED** [review]
Thanks @aemadrid for this contribution! I've left some comments in the code, please let me know what you think.

**Comment by aemadrid** [comment]
Hopefully my last commit addresses your recommendations

**Comment by noteflakes** [comment]
Thanks @aemadrid ! I'll work on updating the README.

**Comment by aemadrid** [comment]
@ciconia thanks! Hopefully this helps the next developer in adding extensions.

**Pull request merged**